man/callflow.1
.TH CALLFLOW 1 "2011/06/05" callflow "callflow - Sequence Diagram Generator"


.SH NAME
callflow - Sequence Diagram Generator

.SH SYNOPSIS
callflow [options] <trace-file>

.SH DESCRIPTION
This is a collection of awk and shell scripts that will take a capture file that can be read by tshark and produce a callflow sequence diagram.  The scripts have been primarily tested with SIP call flows, but should work for other network traffic as well.

.SS Using callflow
With callflow in your path, just type:
  callflow capture-file.cap
.sp 0
(see below for more options)

This will produce a directory named like your capture file in your working directory (eg: if file is capture-file.cap, the directory will be capture-file).
In this, you will find at least:
  - callflow.svg file
  - callflow.png file
  - several html index files
  - graph.html file
  - frames directory with FrameX.html files inside

You may also find (depending on options chosen):
  - callflow.long
  - callflow.short

You can view callflow.svg with the Adobe SVG plugin or Firefox, or you can view index*.html with any web browser.  Both the SVG file and the html file contain links into the frames directory so that you can look at the contents of the full packet frame.  All the frames have been processed to remove the IP headers, which usually aren't that interesting.

.SH OPTIONS
.PP
.I Callflow
accepts the followings options:
.TP
.PD 0
.B -h
Print the short help message.

.TP
.PD 0
.B --help
Print the long help message.

.TP
.PD 0
.B -v
.TP
.PD
.B --version
Print version number.

.TP
.PD 0
.B --archive <archive type>
Create an archive with the files for showing the callflow diagram.  Available archive types are: files, bz2 (bzip2) and zip.  The callflow archive can be used to sent to interested parties.  The archive type \fIfiles\fR just list the files required for the callflow diagram.  It can be used to create an archive with it.

.TP
.PD 0
.B --browser
Launch a browser with the callflow.

.TP
.PD 0
.B --camel-case
Node names are displayed in a high-low alternating pattern. It prevents overlapping of long node names.  By default the node labels are displayed in a horizontal line.  Depending on the on the horizontal width between the nodes and the longest node name, the high-low alternating pattern will be enabled automatically.  This option enforces the high-low alternating pattern for node names.


.TP
.PD 0
.BI --lines-between-nodes "<value>"
Lines between nodes determines the number of lines after which the node labels are repeated.  Value is not restricted nor checked, a sensible value is between 5 and 30.  Default value: 28.

.TP
.PD 0
.BI --filter " filter"
A filter can be used to show only those frames that match the given filter criteria.  This filter is for traces that have been captured with programs such as wireshark, snoop, tcpdump, thsark and the like.  Callflow uses tshark's DISPLAY filter functionality to filter the trace.  More information about tshark's DISPLAY filter can be obtained from the tshark man page. A useful filter for SIP traffic is "sip".

.TP
.PD 0
.B -o
.TP
.PD
.B --list-nodes
List the unique nodes in the trace file (don't forget to redirect output in another file. Example of use: callflow --list-nodes capture.cap > order).  See the description of the file \fIorder\fR in the \fIfiles\fR section.

.TP
.PD 0
.B --no-archive
Disable the creation of and reference to the archive containing the callflow.

.TP
.PD 0
.B --no-loops
Prevent flows from and to the same equipment to be represented.

.TP
.PD 0
.B --no-sdp
Do not show detailed SDP information in the callflow.

.TP
.PD 0
.B --no-time
Do not show the time.

.TP
.PD 0
.B --nodenames <filename>
A file containing the element or node names that belong to the
addresses shown in the callflow.  The file is space separated, the first column
contains the IP address of the node the rest of the line is the node alias.
E.g.:
.sp 0
10.10.20.30 Subscriber A
.sp 0
10.20.100.111 Asterisk
.sp 0
See the description of the file \fInodenames.conf\fR in the \fIfiles\fR section.

.TP
.PD 0
.B --pdf
Additionally provide the callflow in pdf.

.TP
.PD 0
.B -d
.TP
.PD
.B --remove-duplicate-frames
Remove duplicate frames while computing. This option should not be selected on the first run.

.TP
.PD 0
.BI --title " title"
Specify the title of the callflow.  If not specified the file name containing the trace will be used.

.TP
.PD
.B --refresh-cache
Use callflow and callflow.long files from FILE directory.
.sp 0
Note 1: if --refresh-cache and --remove-duplicate-frames are used, only --refresh-cache is executed.
.sp 0
Note 2: this option sould not be selected on the first run.

.TP
.PD 0
.BI --width-between-nodes " width"
Distance between the nodes in final picture.  The distance may vary between 100 and 500.  Default = 400.

.TP
.PD 0
.B --with-sdp
Show detailed SDP information in the callflow (default)


.SS Trace-file
Two trace-file formats are supported; network traces (files that can be read by wireshark, tshark and the like) and Broadsoft Broadworks (SIP) server logs.  In case a Broadworks log is used, the Broadworks server IP address can be specified in callflow's configuration files.


.SS Filtering the diagram
A filter can be used to show only those frames that match the given filter criteria.  This filter is for traces that have been captured with programs such as wireshark, snoop, tcpdump, thsark and the like.  Callflow uses tshark's DISPLAY filter functionality to filter the trace.  More information about tshark's DISPLAY filter can be obtained from the tshark man page. A useful filter for SIP traffic is "sip".


.SS Ordering the diagram
When you know which nodes in your network are logically connected, you can have the diagram ordered in a meaningful way.  Just create a file called order in your current directory or in the setup directory with the order of your network nodes, one per line.  You must use the exact name or IP address that appears in your capture file in the form: <host/ip>:<port>, or use a regular expression that resolves to that name. Using a regular expression can be useful to collapse several nodes in to one logical node.

You can generate an order file with:
  callflow --list-nodes capture-file.cap > order

The format of the order file is as follows:
<node-regex-pattern> [optional-label]

Network nodes that are in your filtered capture file but not in your order file will appear on the right of the diagram, and nodes that are in your order file but not in the filtered capture will not be present in the diagram.


.SS Labeling the diagram
First, you need to create on order file, as above.  Anything that appears after the node name or IP will be used as the label in the diagram.  Otherwise, the node name or IP will be used as the label.


.SS Forcing a node into the diagram
When you would like a node that has no traffic to appear in the diagram, put the string "!f!" as part of its label in the order file. 
The string "!f!" will be edited out of the label.  This can be useful to show the position of a firewall in a trace, or to illustrate a proxy that does not receive traffic.
Make sure that the forced node does not resolve to a regex pattern that another node will match! (ie. don't use ".")

.I Example order file
 10.250.250.101:(1917|5060) Pingtel
 10.250.250.201:5060 VOCAL-UAMS
 10.250.250.201:5070 VOCAL-RS
 10.250.250.201:5065 VOCAL-GWMS
 firewall !f!Firewall
 10.250.250.104:5060 PSTN-GW


.SS Titling the diagram
You can title the diagram by using the --title option. If no --title option is provided, a default title based on the filename will be used.

.SS Removing Duplicate Frames
You can remove duplicate frames from the short text-file with:
callflow --remove-duplicate-frames foo

Duplicates are determined by examining each line in foo, and comparing frames/Frame<#>.html with all previously seen frames.

.SS "Session ID" to Sequence Lines
Callflow colors every frame based on the specific pattern: 'Call-Id:' or 'i:' (for abbreviated SIP messages).


.SH TWEAKING TSHARK PREFERENCE
You can change the level of detail provided in the "Internet Protocol", "User Datagram Protocol", and "Transmission Control Protocol" sections of the detailed frames/Frame*.html pages by tweaking the following entries in your ~/.wireshark/preferences:

ip.ip_summary_in_tree
tcp.tcp_summary_in_tree
udp.udp_summary_in_tree

If any of these items are set to TRUE, then only the summary line will appear in frames/Frame*.html. Otherwise, the gory details will be displayed.


.SH CREATING DIAGRAM MANUALLY
You can manually create a diagram by creating two text files, foo.short and foo.long.

When you run callflow, you can take the output files callflow.short and callflow.long as a starting point.

The first file (.short) contains the trace information, one transaction per line, in the following format:

 <time>|<trace-filename>|<frame #>|<src node>|<src port>|<sessionID>|<dest node>|<dest port>|<protocol>|<summary>|<remark>

.I For example:

 16:07:56.616502||31|172.20.154.66|sip|{1}|172.20.154.92|dsmeter_iatc|SIP/SDP|Request: INVITE sip:68@openims.fr, with session description
 16:07:56.617453||32|172.20.154.92|dsmeter_iatc|{1}|172.20.154.66|sip|SIP|Status: 100 trying -- your call is important to us
 etc.

  / \\
 / ! \\ It is important that there is *no blank lines* in the short text-file.
 -----



If this file is called foo.short, you can create another file called foo.long with longer descriptions of each transaction, in the following format:

 Frame #
 Descriptive information
 goes here

 Frame #
 Description of another frame


.I For example:

 Frame 1
 Via: SIP/2.0/UDP 192.168.1.112:5060
 From: "C7960 (x1201)" <sip:1201@192.168.1.42>;tag=aab70900293102348a-7008
 To: <sip:1020@192.168.1.42>

 Frame 2
 To: <sip:1020@192.168.1.42:5060>;tag=53d20696
 From: "C7960 (x1201)"<sip:1201@192.168.1.42:5060>;tag=aab70900293102348a-7008
 Call-ID: aab70900-2fe3102-3ae-2b27@192.168.1.112

 etc.


To process your text files, type the following:
  callflow -t capture-file.cap


.SH ADDING COMMENTS
You can add a comment to your diagram by adding a line that start with "# " to short text-file. If in the middle of the line there is a " ! " token, then the comment becomes a hyper-link with what precedes the " ! " is the text that is displayed, and what comes after is the link to be followed when a user clicks on the text.

.I For example:

 # Click here for original wireshark capture file ! mycapture.cap
 # This is a comment
 1 Alice 1000 -> Bob 2000 Alice sends Bob an Invite
 2 Bob 2000 -> Alice 1000 Bob sends Alice a 200
 etc.

  / \\
 / ! \\ It is important that there be *no blank lines* in the short text-file.
 -----


.SH Example
.SS Create an order file:
 $ callflow --list-nodes mycapture.cap > order
 (edit order file as needed)

.SS Generate initial diagram:
 $ callflow mycapture.cap

.SS Remove retransmitted SIP packets:
We will now use "mycapture/callflow.short" and "mycapture/callflow.long" text files for further processing:
 $ callflow --remove-duplicate-frames mycapture.cap

.SS Add any desired comments to callflow, then regenerate diagram (repeat as needed):
 $ callflow mycapture.cap


.SH HOW IT WORKS
Well, this is kind of ugly:

    a.  look for the order files
    b.  run tshark twice to get short and long output files
    c.  use long2html.awk to convert long output file into frame*.html files
    d.  get all the unique nodes from the short output file
    e.  order them and make sure all nodes are present
    f.  create the first few lines of the main awk script using makevars.awk
    g.  run the main awk script (callflow.awk) on the short output file:
        I.     create SVG file header and html <map> header
        II.    create labels in SVG file
        III.   for each line, create an arrow in the SVG file and a
               <area> element for the image map
        IV.    write out the SVG and map file footers

    h.  generate callflow.png using inkscape
    i.  generate html index files
    i.  clean up all the /tmp files


.SH "FILES"
.SS callflow.conf
The system-wide configuration file \fI/usr/local/etc/callflow/callflow.conf\fR, and the personal ones \fI$HOME/.callflow/callflow.conf\fR
and \fIcallflow.conf\fR are used by callflow.  If the system-wide configuration file exists, it is read first,
overriding the default settings.  If the personal configuration files exists, they are read next, overriding
any previous values.

.SS order
The file order is expected in the current working directory, and determines the order of the nodes in the callflow.

.SS nodenames.conf
The file nodenames.conf is expected in respectively $HOME/.callflow, the current working directory or the file can be defined with the argument --nodenames.  The last nodenames.conf file (or the file specified with --nodenames) found will be used.

.SH TO DO
It would be much nicer to take the tshark filtered output and build an intermediate XML representation of it.  Then use XSLT to transform it into the SVG file.  This way you could write up illustrative sequence diagrams more easily and make use of more tools that can manipulate XML.

Write port numbers at the ends of each arrow in an unobtrusive fashion.

Automatically find a "best order" for the nodes, possibly based on total arrow length minimization.

Specify output filenames and locations on the command line.


.SH "SEE ALSO"
Callflow project page at \fIhttp://callflow.sourceforge.net\fR


.SH AUTHORS
Man page created by Arnaud Morin <arnaud.morin@gmail.com>.

See README.md file provided with this package to see all project contributors.



debian/changelog
callflow (20091216-1) unstable; urgency=low

  * New release

 -- Arnaud Morin <arnaud.morin@gmail.com>  Wed, 16 Dec 2009 17:56:29 +0100

callflow (20090404-1) unstable; urgency=low

  * Initial release

 -- Arnaud Morin <arnaud.morin@gmail.com>  Sun, 05 Apr 2009 14:42:25 +0100



scripts/callflow.awk


  print "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>";
  print "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.0//EN\" \"http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd\">";

  longNames = "no"

  # A character takes on average between 6-7 pixels (dots).  When 7 is used a lot of 
  # node name space is offered to prevent overlapping node names.  When 6 pixels are used
  # a little overlap occurs at the boundary of activating the camel pattern.  In this
  # case it is up to user to play with --width-between-nodes value to prevent overlapping
  # name labels or use shorter node names.
  maxCharsForName = int ( xHostSpace / 6 )

  for(i=0;i<numHosts;i++) {
    lookup[hosts[i]] = i;
    printf "<!-- lookup['%s'] = %d -->\n", hosts[i], i;
    if ((camelcase == "always") || (length(label[i]) > maxCharsForName))
      longNames = "yes";
  }

  nodes_extra_height = ( int(numLines / yLinesBetweenNodes) + 1 ) * yHostNameSpace
  if (longNames == "yes") {
    # Take the camel pattern into account when the node names are long
    nodes_extra_height *= 2 
  }

  w = (numHosts-1) * xHostSpace + leftMargin + rightMargin;
  h = numLines * yLineSpace + topMargin + bottomMargin + nodes_extra_height;
  yend = h;
  printf "<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"%d\" height=\"%d\" viewBox=\"0 0 %d %d\">\n", w, h, w, h

  # Info for debugging
  debug = 0
  if (debug) {
    printf "<!--\n"
    printf "  DEBUG: == Set Values ==\n"
    printf "  DEBUG: topMargin = %d\n", topMargin
    printf "  DEBUG: bottomMargin = %d\n", bottomMargin
    printf "  DEBUG: rightMargin = %d\n", rightMargin
    printf "  DEBUG: leftMargin = %d\n", leftMargin
    printf "  DEBUG: xHostSpace = %d\n", xHostSpace
    printf "  DEBUG: yLineSpace = %d\n", yLineSpace
    printf "  DEBUG: yLinesBetweenNodes = %d\n", yLinesBetweenNodes
    printf "  DEBUG: \n" 
    printf "  DEBUG: == Calculated Values ==\n"
    printf "  DEBUG: maxCharsForName = %d\n", maxCharsForName;
    printf "  DEBUG: numlines = %d\n", numLines
    printf "  DEBUG: Extra inline node name lines = %d\n", int(numLines / yLinesBetweenNodes)
    printf "  DEBUG: nodes_extra_height = %d\n", nodes_extra_height
    printf "  DEBUG: yHostNameSpace = %d\n", yHostNameSpace
    printf "  DEBUG: Height = %d\n", h
    printf "  DEBUG: Width = %d\n", w
    printf "-->\n\n"

    # The actual size of the drawing can be obtained by executed the following
    # shell script in the data directory.
    # ( grep -E "<polygon" $DESTDIR/callflow.svg |
    #     awk -F\" '{print $2}' | awk -F, '{print $NF}' 
    # 
    #   grep -E "<line|<text" callflow-dir/callflow.svg |
    #     awk '{print $3}' | cut -d\" -f2
    # 
    # ) | sort -n | tail -1
  }

  insertStyleDefs();

  y = 25;

  # With 'leftMargin + ( 1.5 * xHostSpace )' the title is centered above
  # the second column
  printf "<text x=\"%d\" y=\"%d\" class=\"label title-text\">%s</text>\n",
    leftMargin + ( 1.5 * xHostSpace ),
    y,
    title;

  y += yHostNameSpace

  # Take the camel pattern into account in case the node names are long
  if (longNames == "yes") {
    y += yHostNameSpace
  }

  # 1: print the vertical node lines
  print_nodes(y, 1);

  printf "   <map name=\"callflowmap\" id=\"callflowmap\">\n" > imageMap;
}


################ Fuction Definition  ##################

func insertStyleDefs () {
  printf "<defs>\n<style type=\"text/css\"><![CDATA[\n";
  printf " .traceline { stroke-width: 1pt; stroke:black; }\n";
  printf " .pkt-text { color: red; font-family: Trebuchet MS,Helvetica, sans-serif;\n";
  printf "            font-size: 10pt; stroke: none; fill: black;}\n";
  printf " .comment-text { color: black; font-family: Trebuchet MS,Helvetica, sans-serif;\n";
  printf "            font-size: 10pt; stroke: none; fill: green;}\n";
  printf " .host-text { color: black; font-family: Trebuchet MS,Helvetica,sans-serif;\n";
  printf "             font-size: 10pt; stroke:none; fill:blue;}\n";
  printf " .title-text { color: black; font-family: Trebuchet MS,Helvetica,sans-serif;\n";
  printf "              font-size:16pt; stroke:none; fill:black;}\n";
  printf " .label { color: blue; text-anchor: middle ; }\n";
  printf " .arrowhead { stroke-width: 0.5pt;stroke:black; }\n";
  printf " .element-line { stroke-width: 0.25pt; stroke: black; }\n";
  printf " .link { color: blue; font-family: Trebuchet MS,Helvetica, sans-serif;\n";
  printf "		font-size: 9pt; text-decoration: underline; fill: blue;}\n";
  # printf " .session-text { color: red; font-family: Trebuchet MS,Helvetica, sans-serif;\n";
  # printf "		font-size: 9pt; font-weight: bold; fill: red;}\n";
  printf " ]]></style>\n</defs>\n";
}

func arrow(x,y,d,c) {
  
  printf "<polygon points=\"%d,%d %d,%d %d,%d %d,%d\" class=\"arrowhead\" style=\"fill: %s; stroke: %s;\"/>\n",
  x,y,
  x+5*d, y-3,
  x+3*d, y,
  x+5*d, y+3,
  color[c], color[c];
}

func line(x1,x2,y,output, c) {

  if (x1 == x2) {
    printf "<polyline points=\"%d,%d %d,%d %d,%d %d,%d %d,%d %d,%d %d,%d %d,%d\" fill=\"none\" class=\"traceline\" style=\"stroke: %s;\"/>\n",
      x1,y-18+7,
      x1+5,y-18+7,
      x1+12,y-15+7,
      x1+15,y-10+7,
      x1+15,y-5+7,
      x1+12,y-2+7,
      x1+5,y+7,
      x1,y+7,
      color[c];
    arrow(x1,y+7,1,c);

    xtext = x1 + 18;
    
    printf "    <area href=\"frames/Frame%d.html\" coords=\"%d,%d,%d,%d\" alt=\"frame %d\"  onmouseover=\"return getFrame('frames/Frame%d.html');\" onmouseout=\"return nd();\"/>\n", $3, x1, y-yLineSpace+2, x1+15, y+7+1, $3, $3 >> imageMap

  } else if (x1<x2) {
    printf "<line x1=\"%d\" y1=\"%d\" x2=\"%d\" y2=\"%d\" class=\"traceline\" style=\"stroke: %s;\"/>\n", x1, y, x2, y, color[c];
    arrow(x2,y,-1,c);
  
    xtext = x1 + 10;
    
    printf "    <area href=\"frames/Frame%d.html\" coords=\"%d,%d,%d,%d\" alt=\"frame %d\"  onmouseover=\"return getFrame('frames/Frame%d.html');\" onmouseout=\"return nd();\"/>\n", $3, x1, y-yLineSpace+2, x2, y+1, $3, $3 >> imageMap

  } else {
    printf "<line x1=\"%d\" y1=\"%d\" x2=\"%d\" y2=\"%d\" class=\"traceline\" style=\"stroke: %s;\"/>\n", x1, y, x2, y, color[c];
    arrow(x2,y,1,c);
    
    xtext = x2 + 10;
    
    printf "    <area href=\"frames/Frame%d.html\" coords=\"%d,%d,%d,%d\" alt=\"frame %d\"  onmouseover=\"return getFrame('frames/Frame%d.html');\" onmouseout=\"return nd();\"/>\n", $3, x2, y-yLineSpace+2, x1, y+1, $3, $3 >> imageMap
  }
  
  printf "<a href=\"frames/Frame%d.html\" target=\"_blank\">\n", $3;
  printf "<text x=\"%d\" y=\"%d\" class=\"pkt-text\">%s</text>\n", xtext, y-4, output;
  printf "</a>\n";
}

func print_nodes(yPos, first_line) {
  for(i=0;i<numTraces;i++) {

    if (label[i] == "")
      label[i] = hosts[i];

    #  Display the node names in a camel case (low-high) pattern when longNames == yes
    if (longNames == "yes") {
       adjustment = yHostNameSpace*(i%2)
    }

    printf "<text x=\"%d\" y=\"%d\" class=\"label host-text\">%s</text>\n",
      leftMargin+(i*xHostSpace),
      yPos-adjustment,
      label[i];

    # Drawing the lines for the actors only needed once
    if (first_line){
      printf "<line x1=\"%d\" y1=\"%d\" x2=\"%d\" y2=\"%d\" class=\"element-line\" />\n", leftMargin+(i*xHostSpace), yPos+2, leftMargin+(i*xHostSpace), yend;
    }
  }
}

################ Main ##################
{
  if ( NR % yLinesBetweenNodes == 0 ) {

    y += yHostNameSpace

    if ( longNames == "yes" ) {
      y += yHostNameSpace
    } 

    # 0: only print the node names
    print_nodes(y, 0)
  }

  y += yLineSpace

  if ($0 ~ "^#") {

    # The "!" is the link identifier
    LINK = index($0, "!")
    if (LINK == 0) {
      # There is no link, everything behind the first "#" is the comment
      sub("#", "", $0)
      printf("<text x=\"%d\" y=\"%d\" class=\"comment-text\" xml:space=\"preserve\">%s</text>\n", leftMargin + 5, y, $0);

    } else {
      # Make the comment a hyperlink
      # - Everything before the "!" is the link text
      # - Everything behind the "!" is the hyperlink
      output = substr($0, 2, LINK - 2)
      link = substr($0, LINK + 1)

      printf("<text x=\"%d\" y=\"%d\" class=\"link\" xml:space=\"preserve\">%s</text>\n", leftMargin + 5, y, output);
      printf("    <area href=\"%s\" coords=\"%d,%d,%d,%d\"/>\n", link, leftMargin + 5, y-yLineSpace+2, w, y+1) >> imageMap;
    }

  } else {

    if ($6 ~ "{([0-9]+)}") {
      str = $6
      gsub("{", "", str)
      gsub("}", "", str)
    
      colorId = strtonum(str)
      if ( colorId <= colors ) {
        c = colorId
      } else {

        # There are more sessions than configured colors.
        # choose the first color as default...
        c = 0
      }

    } else { c = 0 }

    l1 = sprintf("%s:%s", $4,$5);
    l2 = sprintf("%s:%s", $7,$8);
    for (i=0; i<numHosts; i++) {

      if (l1 ~ hosts[i]) {
        x1 = strtonum(lookup[hosts[i]]);
      }

      if (l2 ~ hosts[i]) {
        x2 = strtonum(lookup[hosts[i]]);
      }
    }

    x1 = x1 * xHostSpace + leftMargin;
    x2 = x2 * xHostSpace + leftMargin;
    if ((x1==x2) && (localLoop == 0)) {
      # Do nothing
    } else {
      # Print the line
      ORS = "";
      printf ("<text x=\"%d\" y=\"%d\" class=\"pkt-text\" xml:space=\"preserve\" ", leftMargin - 10, y)
      printf ("style=\"text-anchor: end;\">%d", $3)
      if ( showTime == "yes" ) {
        printf ("    %s", $1)
      }
      printf ("</text>\n")

      if ((showSDP == "yes") && ($9 ~ "SDP")) {
        output = sprintf("%s: a=%s:%s %s %s", $10, $12, $13, $14, $15)
      } else {
        output = $10
      }

      line(x1, x2, y, output, c);
    }
  }
}

END {
  printf "   </map>\n" >> imageMap;
  printf "</svg>\n";
}

